### PR TITLE
[NSLayoutManager] Complete bindings for NSLayoutManager. Fixes #4740.

### LINKED ISSUE #4740: Multiple LineGlyph bindings are NSLayoutManager are missing due to NSRangePointer usage
I was unable to port Apple's CircleView sample without an ugly manual binding:

```
	public static class LayoutManagerExtensions
	{
		static readonly IntPtr selLineFragmentRectForGlyphAtIndex_EffectiveRange_Handle = ObjCRuntime.Selector.GetHandle ("lineFragmentRectForGlyphAtIndex:effectiveRange:");

		[DllImport ("/usr/lib/libobjc.dylib", EntryPoint = "objc_msgSend_stret")]
		public extern static void CGRect_objc_msgSend_stret_nuint_IntPtr (out CGRect retval, IntPtr receiver, IntPtr selector, nuint arg1, IntPtr arg2);

		public static CGRect LineFragmentRectForGlyphAtIndex (this NSLayoutManager manager, nuint glyphIndex, IntPtr effectiveGlyphRange)
		{
			NSApplication.EnsureUIThread ();
			CGRect_objc_msgSend_stret_nuint_IntPtr (out CGRect ret, manager.Handle, selLineFragmentRectForGlyphAtIndex_EffectiveRange_Handle, glyphIndex, effectiveGlyphRange);
			return ret;
		}
	}
```

as we did not bind a number of glyph APIs that use NSRangePointer. The Swift sample passes null, so could we bind them for now with IntPtr.Zero?

### PR DESCRIPTION
**Pull request by rolfbjarne**
Fix numerous issues with NSLayoutManager[Delegate]:

* The classes are available in both AppKit and UIKit, but the bindings are
  duplicated (unsuccessfully) in both appkit.cs and uikit.cs. So create a new
  xkit.cs that is shared between XI and XM, and put a shared version of the
  bindings there.

* Bind everything that hasn't already been bound (or deprecated by Apple).

* Methods that take a nullable NSRangePointer has been bound with three overloads:

	* A protected overridable (exported) method that uses IntPtr.
	* A public method without the parameter.
	* A public method with the parameter typed as 'ref NSRange'.

	This makes sure the native method can be overridden if needed, while at
	the same time making it possible to call without providing the nullable
	parameter.

* Fix numerous ugly bindings:

	* There's a great nint/nuint confusion for parameters referring to
	  'character index' and 'glyph index'. XI seems to prefer nuint, while XM
	  seems to prefer nint. Standardize on nuint, since that's how Apple
	  created them.

	* Many methods have names than sound like Objective-C. Fix them all,
	  either right away when possible, or for XAMCORE_4_0.

	* Several parameter names have been modified to comply with our naming
	  guidelines (no abbreviations).

Fixes https://github.com/xamarin/xamarin-macios/issues/4740.

### CODE REVIEW

File: src/xkit.cs [hunk 1]
```diff
@@ -0,0 +1,1144 @@
...
+
+		[Export ("ensureLayoutForBoundingRect:inTextContainer:")]
+		void EnsureLayoutForBoundingRect (CGRect bounds, NSTextContainer container);
+
+#if !XAMCORE_4_0
+		[NoiOS][NoTV]
+		[Availability (Deprecated = Platform.Mac_10_11, Message = "Use 'SetGlyphs' instead.")]
+		[Export ("insertGlyph:atGlyphIndex:characterIndex:")]
+		void InsertGlyph (NSGlyph glyph, nint glyphIndex, nint charIndex);
+#endif
```
Comment (chamons) [inline src/xkit.cs]: Should we make this one giant #if block?

File: src/UIKit/NSLayoutManager.cs [hunk 2]
```diff
@@ -82,70 +94,76 @@ partial class NSLayoutManager {
...
+			return GetGlyphRange (charRange);
 		}
 
-		public unsafe NSRange GlyphRangeForCharacterRange (NSRange charRange, ref NSRange actualCharRange)
+		[Obsolete ("Use 'GetGlyphRange' instead.")]
+		public NSRange GlyphRangeForCharacterRange (NSRange charRange, ref NSRange actualCharRange)
 		{
-			fixed (NSRange* ptr = &actualCharRange)
-				return GlyphRangeForCharacterRangeInternal (charRange, (IntPtr) ptr);
+			return GetGlyphRange (charRange);
```
[the remark this answers is not in the document]
Comment (rolfbjarne) [reply src/UIKit/NSLayoutManager.cs]: Ops! Fixed now.

File: src/xkit.cs [hunk 3]
```diff
@@ -0,0 +1,1144 @@
...
+
+		[Mac (10,11)]
+		[Export ("enumerateEnclosingRectsForGlyphRange:withinSelectedGlyphRange:inTextContainer:usingBlock:")]
+		void EnumerateEnclosingRects (NSRange glyphRange, NSRange selectedRange, NSTextContainer textContainer, NSTextLayoutEnumerateEnclosingRects callback);
+
+		[Protected] // Can be overridden
+		[Export ("showCGGlyphs:positions:count:font:matrix:attributes:inContext:")]
+		void ShowGlyphs (IntPtr glyphs, IntPtr positions, nuint glyphCount, NSFont font, CGAffineTransform textMatrix, NSDictionary attributes, [NullAllowed] CGContext graphicsContext);
+
+		[Protected] // "You should never call this method, but you might override it."
```
Comment (spouliot) [inline src/xkit.cs]: Add comment (or similar) to an `[Advice]` attribute.

Comment (spouliot) [reply src/xkit.cs]: fwiw it's the `IntPtr` that made me look into the API... but in this case I can see value in avoiding the conversion cost

File: src/xkit.cs [hunk 4]
```diff
@@ -0,0 +1,1144 @@
...
+		[Protected]
+#else
+		[Sealed][Internal]
+#endif
+		[Export ("lineFragmentUsedRectForGlyphAtIndex:effectiveRange:withoutAdditionalLayout:")]
+		CGRect GetLineFragmentUsedRect (nuint glyphIndex, /* nullable NSRangePointer */ IntPtr effectiveGlyphRange, bool withoutAdditionalLayout);
+
+		[iOS (9,0)]
+		[Wrap ("GetLineFragmentRect (glyphIndex, IntPtr.Zero)")]
+		CGRect GetLineFragmentUsedRect (nuint glyphIndex, bool withoutAdditionalLayout);
```
Comment (spouliot) [inline src/xkit.cs]: should not this wrap `GetLineFragmentUsedRect` instead of `GetLineFragmentRect` ?

File: src/xkit.cs [hunk 5]
```diff
@@ -0,0 +1,1144 @@
...
+#endif
+		[Export ("lineFragmentRectForGlyphAtIndex:effectiveRange:withoutAdditionalLayout:")]
+		CGRect GetLineFragmentRect (nuint glyphIndex, out /* nullable NSRangePointer */ NSRange effectiveGlyphRange, bool withoutAdditionalLayout);
+
+		/* GetLineFragmentUsedRect (NSUInteger, NSRangePointer) */
+		[Protected]
+		[Export ("lineFragmentUsedRectForGlyphAtIndex:effectiveRange:")]
+		CGRect GetLineFragmentUsedRect (nuint glyphIndex, /* nullable NSRangePointer */ IntPtr effectiveGlyphRange);
+
+		[Wrap ("GetLineFragmentRect (glyphIndex, IntPtr.Zero)")]
```
Comment (spouliot) [inline src/xkit.cs]: should not this wrap `GetLineFragmentUsedRect` instead of `GetLineFragmentRect` ?

File: src/xkit.cs (hunk 3 above)
Comment (rolfbjarne) [reply src/xkit.cs]: This is a C-based array, so we generator doesn't support binding it nicely.

Also if the only point to the API is to be able to override it, there's no need for a nicer API that can only be called, not overridden.

**Comment by rolfbjarne** [comment]
> which means the RFC, or at least the `ref NSRange?` does not apply here

The RFC is for both `ref` and `out` parameters (I've updated it to make it clearer).

File: src/xkit.cs (hunk 5 above)
Comment (rolfbjarne) [reply src/xkit.cs]: Fixed (and the previous one too).

**Pull request merged**